libdrgn/program.c

```diff
@@ -1108,6 +1108,140 @@ drgn_program_find_symbol_by_address(struct drgn_program *prog, uint64_t address,
 	return NULL;
 }
 
+DEFINE_VECTOR(symbolp_vector, struct drgn_symbol *);
+
+enum {
+	SEARCH_NAME = (1 << 0),
+	SEARCH_ADDRESS = (1 << 1),
+	SEARCH_ALL = (1 << 2),
+};
+
+struct symbols_search_arg {
+	const char *name;
+	uint64_t address;
+	struct symbolp_vector results;
+	struct drgn_error *err;
+	bool bad_symtabs;
+	unsigned int flags;
+};
+
+static bool symbol_match(struct symbols_search_arg *arg, GElf_Addr addr,
+			 const GElf_Sym *sym, const char *name)
+{
+	if (arg->flags & SEARCH_ALL)
+		return true;
+	if (arg->flags & SEARCH_NAME)
+		if (name && arg->name && strcmp(name, arg->name) == 0)
+			return true;
+	if (arg->flags & SEARCH_ADDRESS)
+		if (arg->address >= addr && arg->address < addr + sym->st_size)
```

Comment:
Style nit: I'd squash these two if statements into one:

```suggestion
	if ((arg->flags & SEARCH_ADDRESS) &&
	    arg->address >= addr && arg->address < addr + sym->st_size)
```